### PR TITLE
Cert expiry check 2

### PR DESCRIPTION
**Pull request by xnull**
## Overview

Description:
Second attempt to bring: cert expiry check

Why should this be merged: 
We need to have a mechanism that allows disabling certificate expiration checks.
Certificates expire in 5 years, if that happens we need somehow disable expiration checks and replace certificates by newer ones.
The way how to do that is by creating '/config/corfu/DISABLE_CERT_EXPIRY_CHECK'. Corfu will check if the file exists and disables expiration checks. 
That's gonna be true for any client also, that's why the fileName by default has that path. 
It allows disabling cert check on all clients at one shot on appliances/on-prem env.

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java [hunk 1]
```diff
@@ -12,86 +14,181 @@
...
-            throw new SSLException(errorMessage, e);
-        } catch (KeyStoreException e) {
-            String errorMessage = "Unable to load trust store " + trustStoreConfig.getTrustStoreFile() + ".";
-            throw new SSLException(errorMessage, e);
+        private final ExecutorService executor = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("TrustStore-watcher-%d")
+                .build());
+
+        private volatile CompletableFuture<TrustManagerContext> trustManagerAsync;
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java:101]: I didn't support the revert of the initial implementation because the fix was trivial and it should have been fixed instead of reverted, but now that it is reverted, maybe its a good time to discuss what this patch tries to achieve, is it performance? and does it really matter if this is just an initialization step and doesn't occur frequently. 


Furthermore, the bug that caused the revert was a deadlock and not starvation, but the git commit message says this patch fixes a starvation issue, are you using the two interchangeably or is there an additional starvation problem that wasn't surfaced with the deadlock bug?

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java:101]: I think the handshake can take more than the 500 ms (i.e., default value) irrespective of this patch therefore that issue should be addressed by tuning the timeout value.

From what I understand, this patch optimizes the specific cases where two or more connections are initialized within a 3 second window, because the trust manager is loaded once every 3 seconds, then connections within that time frame can share the same store without needing to load it from the file system. 

I think this patch isn't necessary for the following reasons:
1. The client connections are long running and few
2. The optimization uses async APIs for what is inherently blocking logic
3. The file system latency isn't really decoupled from the handshake because even if the value is cached the file system is still read via TrustStoreConfig::isCertExpiryCheckEnabled()

In my opinion the complexity/optimization trade-off isn't worth it because the optimize scenario is infrequent and unlikely to happen. For example, look at this method which just reads a file `TlsUtils::openCertStore` , it's really convoluted because it introduces threading/coordination of what is simply sequential blocking code 

```java

public static CompletableFuture<KeyStore> openCertStore(CertStoreConfig certStoreCfg) {
        return CompletableFuture
                //validation step
                .runAsync(certStoreCfg::checkCertStoreExists, EXECUTOR)
                .thenCompose(empty -> getCertStorePassword(certStoreCfg.getPasswordFile()))
                .thenApplyAsync(certStorePassword -> {
                    KeyStore keyStore = getKeyStoreInstance();

                    Path certStore = certStoreCfg.getCertStore();
                    FileInputStream inputStream = null;
                    try {
                        inputStream = new FileInputStream(certStore.toFile());
                        keyStore.load(inputStream, certStorePassword.toCharArray());
                        return keyStore;
                    } catch (IOException e) {
                        String errorMessage = String.format("Unable to read key store file %s.", certStore);
                        throw new IllegalStateException(errorMessage, e);
                    } catch (CertificateException e) {
                        String errorMessage = String.format("Unable to read certificates in key store file %s.", certStore);
                        throw new IllegalStateException(errorMessage, e);
                    } catch (NoSuchAlgorithmException e) {
                        String errorMessage = String.format("No support for algorithm [%s] in key store %s.",
                                KeyStore.getDefaultType(), certStore);
                        throw new IllegalStateException(errorMessage, e);
                    } finally {
                        Consumer<IOException> emptyExceptionHandler = ex -> {
                        };
                        IOUtils.closeQuietly(inputStream, emptyExceptionHandler);
                    }
                }, EXECUTOR);
    }

```

If this method wasn't using the asynchronous completable future APIs the initial bug - deadlock - would have not happened.

In conclusion, I don't think this is a good optimization because its not general and the implementation is unnecessarily complicated. If you still want to have this optimization, then you can have a cleaner implementation with a Guava loading cache that will expire the cache every 3 seconds, that will eliminate most of the async  code.

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java:101]: The patch tries to achieve two things: Cache a file and read the file asynchronously. I think you still pursue the caching part, but without the asynchronous code because of the aforementioned reasoning. Notice that not only can the file be cached, but you can also cache the File.exists check as well.

File: runtime/src/main/java/org/corfudb/security/tls/TlsUtils.java [hunk 2]
```diff
@@ -151,21 +153,35 @@ public Path getCertStore() {
 
         @AllArgsConstructor
         @Getter
+        @ToString
         class TrustStoreConfig implements CertStoreConfig {
+            public static final Path DEFAULT_DISABLE_CERT_EXPIRY_CHECK_FILE = Paths.get(
+                    "/", "config", "corfu", "DISABLE_CERT_EXPIRY_CHECK"
+            );
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/security/tls/TlsUtils.java:160]: Rather than hard-coding this, can't you get the root data directory from the `-l` parameter? This approach won't work for non-config Corfu.

File: runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java [hunk 3]
```diff
@@ -35,16 +34,43 @@ public ReloadableTrustManager(TrustStoreConfig trustStoreConfig) throws SSLExcep
...
     @Override
     public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        if (isCertExpiryCheckEnabled()) {
+            for (X509Certificate cert : chain) {
+                cert.checkValidity();
+            }
+        } else {
+            logCertExpiryCheck();
+        }
+
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java:58]: Can you consolidate this block and the one above into the same function -- i.e. `checkCertValidity`?

Comment (PavelZaytsev) [reply runtime/src/main/java/org/corfudb/security/tls/ReloadableTrustManager.java:58]: +1

**Review by PavelZaytsev — APPROVED** [review]

File: runtime/src/main/java/org/corfudb/security/tls/TlsUtils.java [hunk 4]
```diff
@@ -151,21 +153,35 @@ public Path getCertStore() {
...
             }
 
             @Override
             public Path getCertStore() {
                 return trustStoreFile;
             }
+
+            public boolean isCertExpiryCheckEnabled() {
+                return !Files.exists(disableCertExpiryCheckFile);
+            }
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/security/tls/TlsUtils.java:184]: Is there a requirement for this file-based argument passing? In other words, how come the flag `--disable-cert-expiry-check-file` is not enough?

**Review by vjeko — APPROVED** [review]
LGTM. Also, it would be great if you could follow standardized commit message guidelines: https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53 You should wrap any git messages to 72 characters.
